### PR TITLE
[std]: Use XXHash64 instead FNV-1a

### PR DESCRIPTION
**Pull request by MaxGraey**
This experiment with XXHash 64-bit hashes

See 32-bit here: https://github.com/AssemblyScript/assemblyscript/pull/1580

#### FNV-1a:

- Poor uniform distribution. Only up to 14 bits can be used for the hash table (up to 2 ^ 14 hash table slots) without degradation
- Bad avalanche effect
- Speed: 4.5 cyclers per byte

#### XXHash:

- Excellent uniform distribution
- Near-perfect avalanche effect
- Speed: ~1.7 cyclers per byte

- [x] I've read the contributing guidelines

### CODE REVIEW

File: std/assembly/util/hash.ts [hunk 1]
```diff
@@ -10,63 +8,137 @@ export function HASH<T>(key: T): u32 {
...
 }
 
-function hashStr(key: string): u32 {
-  var v = FNV_OFFSET;
-  if (key !== null) {
-    for (let i: usize = 0, k: usize = key.length << 1; i < k; ++i) {
-      v = (v ^ <u32>load<u8>(changetype<usize>(key) + i)) * FNV_PRIME;
+// @ts-ignore: decorator
+@inline
+function hashStr(key: string): u64 {
```
Comment (dcodeIO) [inline std/assembly/util/hash.ts:75]: Do you have a reference for the code in this function?

Comment (MaxGraey) [reply std/assembly/util/hash.ts:75]: It's loosely based on this: https://github.com/Cyan4973/xxHash/blob/dev/xxhash.h

File: std/assembly/map.ts [hunk 2]
```diff
@@ -58,7 +58,7 @@ export class Map<K,V> {
 
   // buckets referencing their respective first entry, usize[bucketsMask + 1]
   private buckets: ArrayBuffer = new ArrayBuffer(INITIAL_CAPACITY * <i32>BUCKET_SIZE);
-  private bucketsMask: u32 = INITIAL_CAPACITY - 1;
+  private bucketsMask: u64 = INITIAL_CAPACITY - 1;
```
[the remark this answers is not in the document]
Comment (MaxGraey) [reply std/assembly/map.ts:61]: I tied just truncate `u64` result to `u32` inside hashes but it looks less optimal with more u64 -> u32 conversions

[the remark this answers is not in the document]
Comment (dcodeIO) [reply std/assembly/map.ts:61]: I think this makes me a bit in favor of the 32-bit variant for now. Would you agree with a plan like, let's use 32-bit now, and once we have Memory64 support, revisit 64-bit? Iirc that's about what you suggested earlier as well (sorry).

[the remark this answers is not in the document]
Comment (MaxGraey) [reply std/assembly/map.ts:61]: yeah, I'm ok with that. Will sync XXH32

**Pull request closed without merging**